File: tools/pack.js
const fs = require('fs-extra');
const path = require('path');

const childProcess = require('child_process');

const root = path.join(__dirname, '..');

if (! fs.existsSync(path.join(root, 'pbo'))) {
    fs.mkdirSync(path.join(root, 'pbo'));
}

if (! fs.existsSync(path.join(root, 'tmp'))) {
    fs.mkdirSync(path.join(root, 'tmp'));
}

const TEMPLATE_FOLDER = 'ZGM-17_Achilles_Blufor.Altis';
const TEMPLATE_IMAGE_NAME = 'achilles_blufor';
const TEMPLATE_EXT_MISSION_NAME = 'Zeus 16+1 Achilles Altis';
const TEMPLATE_SQM_MISSION_NAME = 'Zeus 16+1 Achilles Altis (BLUFOR)';
const TEMPLATE_SIDE = 'WEST';
const TEMPLATE_SOLDIER = 'B_Soldier_Universal_F';

const FOLDER_PREFIX = 'ZGM-17_Achilles';
const IMAGE_NAME_PREFIX = 'achilles';
const NAME_PREFIX = 'Zeus 16+1 Achilles';

// Define side related array params
const sideNames = ['blufor', 'opfor', 'greenfor'];
const sides = ['WEST', 'EAST', 'GUER'];
const soldiers = ['B_Soldier_Universal_F', 'O_Soldier_Universal_F', 'I_Soldier_Universal_F'];

const whitelist = require('../mapWhitelist.json');

let mapKeys = [];
let mapNames = [];

for (const mapKey in whitelist) {
    if (whitelist.hasOwnProperty(mapKey)) {
        const mapName = whitelist[mapKey];
        mapKeys.push(mapKey);
        mapNames.push(mapName);
    }
}

fs.removeSync(path.join(root, 'missionWhitelist.dat'));

/**
 * @param {number} side Side Index
 * @param {number} map Map Index
 */
function generateMission(side, map) {
    const sideName = sideNames[side];
    const capitalizedSide = sideName.charAt(0).toUpperCase() + sideName.slice(1);

    const imageName = `${IMAGE_NAME_PREFIX}_${sideNames[side]}`;
    const extMissionName = `${NAME_PREFIX} ${mapNames[map]}`;
    const sqmMissionName = `${extMissionName} (${sideNames[side].toUpperCase()})`;

    const folderName = `${FOLDER_PREFIX}_${capitalizedSide}.${mapKeys[map]}`;

    fs.removeSync(path.join(root, 'tmp', folderName));
    fs.copySync(path.join(root, 'src', TEMPLATE_FOLDER), path.join(root, 'tmp', folderName));

    console.log(`Generating ${folderName}`);
    
    if (folderName !== TEMPLATE_FOLDER) {
        fs.removeSync(path.join(root, 'tmp', folderName, 'images'));
        fs.copySync(path.join(root, 'src', 'images', `${imageName}.paa`), path.join(root, 'tmp', folderName, 'images', `${imageName}.paa`));
    
        let descriptionExt = fs.readFileSync(path.join(root, 'tmp', folderName, 'description.ext')).toString(); 

        descriptionExt = descriptionExt.replace(new RegExp(TEMPLATE_EXT_MISSION_NAME.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1'), 'g'), extMissionName);
        descriptionExt = descriptionExt.replace(new RegExp(TEMPLATE_IMAGE_NAME.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1'), 'g'), imageName);

        fs.writeFileSync(path.join(root, 'tmp', folderName, 'description.ext'), descriptionExt);
        
        let missionSqm = fs.readFileSync(path.join(root, 'tmp', folderName, 'mission.sqm')).toString();

        missionSqm = missionSqm.replace(new RegExp(TEMPLATE_SQM_MISSION_NAME.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1'), 'g'), sqmMissionName);
        missionSqm = missionSqm.replace(new RegExp(TEMPLATE_SOLDIER.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1'), 'g'), soldiers[side]);
        missionSqm = missionSqm.replace(new RegExp(TEMPLATE_SIDE.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1'), 'g'), sides[side]);

        fs.writeFileSync(path.join(root, 'tmp', folderName, 'mission.sqm'), missionSqm);
    }
    
    childProcess.spawn('armake', ['build', '-p', '-w', 'unquoted-string', path.join(root, 'tmp', folderName), path.join(root, 'pbo', `${folderName}.pbo`)]);

    fs.appendFileSync(path.join(root, 'missionWhitelist.dat'), `"${folderName}",\n`);
}

for (let i = 0; i < sides.length; i++) {
    for (let j = 0; j < mapKeys.length; j++) {
        generateMission(i, j);        
    }
}

const datFile = fs.readFileSync(path.join(root, 'missionWhitelist.dat')).toString();
fs.writeFileSync(path.join(root, 'missionWhitelist.dat'), datFile.substring(0, datFile.length - 2));

fs.removeSync(path.join(root, 'tmp'));

File: tools/pack.py
#!/usr/bin/env python3

import json
import os
import platform
import re
import shutil
import subprocess
import sys

FOLDER_PREFIX = 'ZGM-21_Zeus_Enhanced'
IMAGE_NAME_PREFIX = 'zen'
NAME_PREFIX = 'Zeus 20+1 Zeus Enhanced'

TEMPLATE_FOLDER = FOLDER_PREFIX + '_Blufor.Altis'
TEMPLATE_IMAGE_NAME = IMAGE_NAME_PREFIX + '_blufor'
TEMPLATE_EXT_MISSION_NAME = NAME_PREFIX + ' Altis'
TEMPLATE_SQM_MISSION_NAME = NAME_PREFIX + ' Altis (BLUFOR)'
TEMPLATE_SIDE = 'WEST'
TEMPLATE_SOLDIER = 'B_Soldier_F'

SIDE_NAMES = ['blufor', 'opfor', 'greenfor']
SIDES = ['WEST', 'EAST', 'GUER']
SOLDIERS = ['B_Soldier_F', 'O_Soldier_F', 'I_Soldier_F']

map_keys = []
map_names = []


def get_map_keys_and_names():
    with open('mapWhitelist.json') as f:
        whitelist = json.load(f)

        for map_key in whitelist:
            map_keys.append(map_key)
            map_names.append(whitelist[map_key])


def generate_mission(armake: str, side_index: int, map_index: int):
    side = SIDES[side_index]
    side_name = SIDE_NAMES[side_index]
    side_capitalized = side_name.capitalize()
    side_upper = side_name.upper()

    map_key = map_keys[map_index]
    map_name = map_names[map_index]

    image_name = f'{IMAGE_NAME_PREFIX}_{side_name}'
    ext_mission_name = f'{NAME_PREFIX} {map_name}'
    sqm_mission_name = f'{ext_mission_name} ({side_upper})'

    soldier = SOLDIERS[side_index]

    pbo_name = f'{FOLDER_PREFIX}_{side_capitalized}.{map_key}'

    dir_tmp = os.path.join('./tmp', pbo_name)
    dir_src = os.path.join('./src')

    print('Generating {}'.format(pbo_name))

    if pbo_name != TEMPLATE_FOLDER:
        # Copy the template
        shutil.copytree(os.path.join(dir_src, TEMPLATE_FOLDER), dir_tmp)

        # Change the image to the right one
        os.remove(os.path.join(dir_tmp, 'images',
                               f'{TEMPLATE_IMAGE_NAME}.paa'))
        shutil.copy(os.path.join(dir_src, 'images', f'{image_name}.paa'),
                    os.path.join(dir_tmp, 'images', f'{image_name}.paa'))

        # Update description.ext
        with open(os.path.join(dir_tmp, 'description.ext'), 'r+') as f:
            content = f.read()

            content = re.sub(re.escape(TEMPLATE_EXT_MISSION_NAME),
                             ext_mission_name, content)
            content = re.sub(TEMPLATE_IMAGE_NAME, image_name, content)

            f.seek(0)
            f.write(content)

        # Update mission.sqm
        with open(os.path.join(dir_tmp, 'mission.sqm'), 'r+') as f:
            content = f.read()

            content = re.sub(re.escape(TEMPLATE_SQM_MISSION_NAME),
                             sqm_mission_name, content)
            content = re.sub(TEMPLATE_SOLDIER, soldier, content)
            content = re.sub(TEMPLATE_SIDE, side, content)

            f.seek(0)
            f.write(content)
    else:
        shutil.copytree(os.path.join(dir_src, pbo_name), dir_tmp)

    # Append to the mission whitelist
    try:
        with open('missionWhitelist.dat', 'a') as f:
            comma = ''

            if len(SIDES) != side_index and len(map_keys) != map_index:
                comma = ','

            f.write(f'"{pbo_name}"{comma}\n')
    except FileNotFoundError:
        with open('missionWhitelist.dat', 'w') as f:
            f.write(f'"{pbo_name}",\n')

    subprocess.call([armake, 'build', '-p', dir_tmp,
                     os.path.join('dist', f'{pbo_name}.pbo')])


def main():
    root_dir = '..'
    if os.path.exists('src'):
        root_dir = '.'

    os.chdir(root_dir)

    get_map_keys_and_names()

    # Clear the tmp dir
    if os.path.isdir('./tmp'):
        shutil.rmtree('./tmp')
    os.mkdir('./tmp')

    # Clear the dist dir
    if os.path.isdir('./dist'):
        shutil.rmtree('./dist')
    os.mkdir('./dist')

    if os.path.isfile('./missionWhitelist.dat'):
        os.remove('./missionWhitelist.dat')

    os_platform = platform.system()

    if os_platform == 'Windows':
        armake = './tools/armake.exe'
    elif os_platform == 'Linux':
        armake = './tools/armake'
    elif os_platform.startswith('CYGWIN'):
        armake = './tools/armake.exe'
    else:
        raise OSError('Unable to build the PBOs for your platform.')

    # Generate all the pbos into a tmp folder
    for side_index in range(len(SIDES)):
        for map_index in range(len(map_keys)):
            generate_mission(armake, side_index, map_index)

    shutil.rmtree('./tmp')


if __name__ == '__main__':
    sys.exit(main())
